src/test/java/jenkins/plugins/git/AbstractGitSCMSourceRetrieveHeadsTest.java

```diff
@@ -0,0 +1,129 @@
+package jenkins.plugins.git;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+
+import hudson.FilePath;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadObserver;
+import org.eclipse.jgit.transport.RefSpec;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+/**
+ * Tests for {@link AbstractGitSCMSource}
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(Git.class)
+public class AbstractGitSCMSourceRetrieveHeadsTest {
+
+    public static final String EXPECTED_GIT_EXE = "git-custom";
+
+    private AbstractGitSCMSource gitSCMSource;
+
+    @Before
+    public void setup() throws Exception {
+        // Mock GitTool
+        GitTool mockedTool = PowerMockito.mock(GitTool.class, Mockito.RETURNS_DEFAULTS);
+        PowerMockito.doReturn(EXPECTED_GIT_EXE).when(mockedTool).getGitExe();
+
+        // Mock git implementation
+        Git git = Mockito.mock(Git.class, Mockito.CALLS_REAL_METHODS);
+        PowerMockito.doThrow(new GitToolSpecified()).when(git).using(EXPECTED_GIT_EXE);
+        PowerMockito.doThrow(new GitToolNotSpecified()).when(git).getClient();
+        PowerMockito.doReturn(git).when(git).in(Mockito.any(File.class));
+        PowerMockito.doReturn(git).when(git).in(Mockito.any(FilePath.class));
+
+        // mock static factory to return our git mock
+        PowerMockito.mockStatic(Git.class, Mockito.CALLS_REAL_METHODS);
+        PowerMockito.doReturn(git).when(Git.class, "with", Mockito.any(), Mockito.any());
+
+        // Partial mock our AbstractGitSCMSourceImpl
+        gitSCMSource = PowerMockito.spy(new AbstractGitSCMSourceImpl());
```

Comment:
Some of the methods on AbstractGitSCMSourceImpl are called internally. This would mean adding return values to the mock (default return values are not suitable -- e.g. mock will return null instead of empty String). I figured it's more readable this way than adding bunch of mock calls.